mne/preprocessing/_regress.py

```diff
@@ -53,35 +74,244 @@ def regress_artifact(inst, picks=None, picks_artifact='eog', betas=None,
-        raise ValueError('picks_artifact cannot be contained in picks')
-    inst = inst.copy() if copy else inst
-    artifact_data = inst._data[..., picks_artifact, :]
-    ref_data = artifact_data - np.mean(artifact_data, -1, keepdims=True)
-    if ref_data.ndim == 3:
-        ref_data = ref_data.transpose(1, 0, 2).reshape(len(picks_artifact), -1)
-    cov = np.dot(ref_data, ref_data.T)
-    # process each one separately to reduce memory load
-    betas_shape = (len(picks), len(picks_artifact))
     if betas is None:
-        betas = np.empty(betas_shape)
-        estimate = True
+        model = EOGRegression(picks=picks, exclude=exclude,
+                              picks_artifact=picks_artifact, taa=taa,
+                              proj=proj)
+        model.fit(inst)
     else:
-        estimate = False
-        betas = np.asarray(betas, dtype=float)
-        _check_option('betas.shape', betas.shape, (betas_shape,))
-    for pi, pick in enumerate(picks):
-        this_data = inst._data[..., pick, :]  # view
-        orig_shape = this_data.shape
-        if estimate:
-            # subtract mean over time from every trial/channel
+        # Create an EOGRegression object and load the given betas into it.
+        picks = _picks_to_idx(inst.info, picks, exclude=exclude, none='data')
+        picks_artifact = _picks_to_idx(inst.info, picks_artifact)
+        want_betas_shape = (len(picks), len(picks_artifact))
+        _check_option('betas.shape', betas.shape, (want_betas_shape,))
+        model = EOGRegression(picks, picks_artifact, taa=taa, proj=proj)
+        model.info = pick_info(inst.info, picks)
+        model.coef_ = betas
+        model._picks = model.info['ch_names']
+        model._exclude = exclude
+        model._picks = model.info['ch_names']
+        model._picks_artifact = [inst.ch_names[ch] for ch in picks_artifact]
+    return model.apply(inst, copy=copy), model.coef_
+
+
+@fill_doc
+class EOGRegression():
+    """Remove EOG artifact signals from other channels by regression.
+
+    Employs linear regression to remove signals captured by some channels,
+    typically EOG, but it also works with ECG from other channels, as described
+    in :footcite:`GrattonEtAl1983`. You can also chose to fit the regression
+    coefficients on evoked blink/saccade data and then apply them to continuous
+    data, as described in :footcite:`CroftBarry2000`.
+
+    Parameters
+    ----------
+    %(picks_good_data)s
+    exclude : list | 'bads'
+        List of channels to exclude from the regression, only used when picking
+        based on types (e.g., exclude="bads" when picks="meg").
+        Specify ``'bads'`` (the default) to exclude all channels marked as bad.
+    picks_artifact : array-like | str
+        Channel picks to use as predictor/explanatory variables capturing
+        the artifact of interest (default is "eog").
+    taa : bool
+        Whether to apply TAA correction as described in equation 9 of
```

Comment:
TAA?

Reply:
An approximation adjustment factor promoted by Croft & Barry (2000). I put it there to follow the paper to the letter. In practice, I don't think I would turn this on. So I removed it.